core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java

```diff
@@ -455,14 +466,13 @@ public String resolveIndexRouting(@Nullable String routing, String aliasOrIndex)
             }
             throw new IllegalArgumentException("Alias [" + aliasOrIndex + "] has more than one index associated with it [" + Arrays.toString(indexNames) + "], can't execute a single index op");
         }
-        AliasMetaData aliasMd = alias.getFirstAliasMetaData();
-        if (aliasMd.indexRouting() != null) {
-            if (routing != null) {
-                if (!routing.equals(aliasMd.indexRouting())) {
-                    throw new IllegalArgumentException("Alias [" + aliasOrIndex + "] has index routing associated with it [" + aliasMd.indexRouting() + "], and was provided with routing value [" + routing + "], rejecting operation");
-                }
+        if (routing == null) {
+            AliasMetaData aliasMd = alias.getFirstAliasMetaData();
+            if (aliasMd != null) {
+                routing = aliasMd.indexRouting();
+            } else if (parent != null) {
+                routing = parent;
             }
```

Comment:
can we keep the `if (!routing.equals(aliasMd.indexRouting())) { throw XXX }` in case both a routing key and an alias routing are set? The alias routing should always win?